adapters/beintoo/beintoo.go

```diff
@@ -0,0 +1,246 @@
+		addHeaderIfNonEmpty(headers, "X-Forwarded-For", request.Device.IP)
+		addHeaderIfNonEmpty(headers, "Accept-Language", request.Device.Language)
+		if request.Device.DNT != nil {
+			addHeaderIfNonEmpty(headers, "DNT", strconv.Itoa(int(*request.Device.DNT)))
+		}
+	}
+	if request.Site != nil {
+		addHeaderIfNonEmpty(headers, "Referer", request.Site.Page)
+	}
+
+	url := buildEndpoint(a.endpoint, a.testing, request.TMax)
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     url,
+		Body:    data,
+		Headers: headers,
+	}}, errs
+}
+
+func unpackImpExt(imp *openrtb.Imp) (*openrtb_ext.ExtImpBeintoo, error) {
+	var bidderExt adapters.ExtImpBidder
+	if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: err.Error(),
+		}
+	}
+
+	var BeintooExt openrtb_ext.ExtImpBeintoo
+	if err := json.Unmarshal(bidderExt.Bidder, &BeintooExt); err != nil {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("ignoring imp id=%s, invalid ImpExt", imp.ID),
+		}
+	}
+
+	tagIDValidation, err := strconv.ParseInt(BeintooExt.TagID, 10, 64)
+	if err != nil || tagIDValidation == 0 {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("ignoring imp id=%s, invalid tagid must be a String of numbers", imp.ID),
+		}
+	}
+
+	if BeintooExt.TagID == "" {
+		return nil, &errortypes.BadInput{
+			Message: fmt.Sprintf("Ignoring imp id=%s, no tagid present", imp.ID),
+		}
+	}
+
+	return &BeintooExt, nil
+}
+
+func buildImpBanner(imp *openrtb.Imp) error {
+	imp.Ext = nil
+
+	if imp.Banner == nil {
+		return &errortypes.BadInput{
+			Message: fmt.Sprintf("Request needs to include a Banner object"),
+		}
+	}
+
+	bannerCopy := *imp.Banner
+	banner := &bannerCopy
+
+	if banner.W == nil && banner.H == nil {
+		if len(banner.Format) == 0 {
+			return &errortypes.BadInput{
+				Message: fmt.Sprintf("Need at least one size to build request"),
+			}
+		}
+		format := banner.Format[0]
+		banner.Format = banner.Format[1:]
+		banner.W = &format.W
+		banner.H = &format.H
+		imp.Banner = banner
+	}
+
+	return nil
+}
+
+// Add Beintoo required properties to Imp object
+func addImpProps(imp *openrtb.Imp, secure *int8, BeintooExt *openrtb_ext.ExtImpBeintoo) {
+	imp.TagID = BeintooExt.TagID
+	imp.Secure = secure
+
+	if BeintooExt.BidFloor != "" {
+		bidFloor, err := strconv.ParseFloat(BeintooExt.BidFloor, 64)
+		if err != nil {
+			bidFloor = 0
+		}
+
+		if bidFloor > 0 {
+			imp.BidFloor = bidFloor
+			imp.BidFloorCur = "USD"
+		}
+	}
+
+	return
+}
+
+// Adding header fields to request header
+func addHeaderIfNonEmpty(headers http.Header, headerName string, headerValue string) {
+	if len(headerValue) > 0 {
+		headers.Add(headerName, headerValue)
+	}
+}
+
+// Handle request errors and formatting to be sent to Beintoo
+func preprocess(request *openrtb.BidRequest) []error {
+	impsCount := len(request.Imp)
+	errors := make([]error, 0, impsCount)
+	resImps := make([]openrtb.Imp, 0, impsCount)
```

Comment:
Small optimization. No need to create `impsCount` since `len(request.Imp)` only reads a lenght from the slice object in O(1) time
```
168   func preprocess(request *openrtb.BidRequest) []error {
169 -     impsCount := len(request.Imp)
170 -     errors := make([]error, 0, impsCount)
171 -     resImps := make([]openrtb.Imp, 0, impsCount)
    +     errors := make([]error, 0, len(request.Imp))
    +     resImps := make([]openrtb.Imp, 0, len(request.Imp))
172       secure := int8(0)
adapters/beintoo/beintoo.go
```